test_plans/weak_object.asciidoc

```diff
@@ -0,0 +1,118 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for weak object
+
+This is a test plan for the APIs described in
+https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_weak_object.asciidoc[SYCL_EXT_ONEAPI_WEAK_OBJECT.asciidoc]
+
+== Testing scope
+
+=== Device coverage
+
+All of the tests described below are performed only on the default device that
+is selected on the CTS command line.
+
+=== Feature test macro
+
+All of the tests should use `#ifdef SYCL_EXT_ONEAPI_WEAK_OBJECT` so they can be skipped
+if feature is not supported.
+
+=== Type coverage
+All of the tests for `weak_object` described below are performed using each of the following `typename SyclObject`:
+
+* `accessor<int>` +
+* `buffer<int>` +
+* `context` +
+* `device` +
+* `event` +
+* `host_accessor<int>` +
+* `kernel` +
+* `kernel_id` +
+* `kernel_bundle<bundle_state::executable>` +
+* `local_accessor<int>` +
+* `platform` +
+* `queue` +
+* `stream` +
+
+== Objects creation
+
+All of the tests for `weak_object` described below using empty or non-empty objects:
+
+* Empty `weak_object` created by using `constexpr weak_object()` constructor
+
+* Non-empty `weak_object` created by using `weak_object(const SyclObject &SYCLObj)` constructor,
+where `SyclObject` was created using one of the following:
+    ** Default class constructor
+    ** Sample data constructor
+    ** `get_<typenmane>()` member functions from other objects. For example `device = queue.get_device()`
+
+== Tests
+
+=== Constructors
+
+Create instances of `weak_object` using: +
+`constexpr weak_object()` +
+`weak_object(const SyclObject &SYCLObj)` +
+`weak_object(const weak_object &Other)` +
+`weak_object(weak_object &&Other)` +
+
+* Check that `object_type` is the same type with `SyclObject` using `std::is_same_v`
+* Check that `try_lock()` returns same `SyclObject` with same value and type using `==` and `std::is_same_v` respectively for non-empty and `std::nullopt` for empty objects
+* Check that `expired()` returns `true` for empty and `false` for non-empty objects
+
+=== Copy and move assignment operators
+
+Use the existing instances of `weak_object` from the previous section other than empty object to copy or move them into other objects +
+`weak_object &operator=(const SyclObject &SYCLObj)` +
+`weak_object &operator=(const weak_object &Other)` +
+`weak_object &operator=(weak_object &&Other)`
+
+* Check that resulting object are the same with initial object using `std::is_same_v`
+* Check that `try_lock()` member function returns same results both in initial and new objects using `==` operator
+* Check that `expired()` returns `false`
+
+=== Reset, swap and expired member functions
```

Comment:
What you have here is correct, but it seems like it would be easier to reuse the tests you define above to test `swap`.  Something like:

> Using the objects created in the "Constructors" section, do the following:
>
> * Create an empty `weak_object` of the same type by using the default constructor.
> * Call `swap` on that object, passing the object from the "Constructors" section.
> * Perform the same tests described above in the "Constructors" section.

Reply:
Same with operators=